cloudinit/cmd/main.py

```diff
@@ -987,7 +987,10 @@ def main(sysv_args=None):
 
     if sysv_args:
         # Only load subparsers if subcommand is specified to avoid load cost
-        subcommand = sysv_args[0]
+        subcommand = next(
```

Comment:
One problem with this solution is that the top-level command supports a `--file` parameter like the following `cloud-init --debug --file /my/path.yaml` which will cause us to claim the subcommand  as `/my/path.yaml`.

I think this top-level --file key is really only appropriate for the following subcommands (init, modules and single).

As such, I think it should only be declared on the specific subcommand parser.
Then our top-level --help will be better able to all subcommands and the `next` handling will hold for simple -- optional parameters.

Here's a diff that pushes the applicable --files optional param down into the 3 subcommands which allow it
```diff
diff --git a/cloudinit/cmd/main.py b/cloudinit/cmd/main.py
index d4bf38243..5bce6917b 100755
--- a/cloudinit/cmd/main.py
+++ b/cloudinit/cmd/main.py
@@ -860,14 +860,6 @@ def main(sysv_args=None):
         version="%(prog)s " + (version.version_string()),
         help="Show program's version number and exit.",
     )
-    parser.add_argument(
-        "--file",
-        "-f",
-        action="append",
-        dest="files",
-        help="Use additional yaml configuration files.",
-        type=argparse.FileType("rb"),
-    )
     parser.add_argument(
         "--debug",
         "-d",
@@ -901,6 +893,14 @@ def main(sysv_args=None):
         help="Start in local mode (default: %(default)s).",
         default=False,
     )
+    parser_init.add_argument(
+        "--file",
+        "-f",
+        action="append",
+        dest="files",
+        help="Use additional yaml configuration files.",
+        type=argparse.FileType("rb"),
+    )
     # This is used so that we can know which action is selected +
     # the functor to use to run this subcommand
     parser_init.set_defaults(action=("init", main_init))
@@ -917,6 +917,14 @@ def main(sysv_args=None):
         default="config",
         choices=("init", "config", "final"),
     )
+    parser_mod.add_argument(
+        "--file",
+        "-f",
+        action="append",
+        dest="files",
+        help="Use additional yaml configuration files.",
+        type=argparse.FileType("rb"),
+    )
     parser_mod.set_defaults(action=("modules", main_modules))
 
     # This subcommand allows you to run a single module
@@ -949,6 +957,14 @@ def main(sysv_args=None):
         metavar="argument",
         help="Any additional arguments to pass to this module.",
     )
+    parser_single.add_argument(
+        "--file",
+        "-f",
+        action="append",
+        dest="files",
+        help="Use additional yaml configuration files.",
+        type=argparse.FileType("rb"),
+    )
     parser_single.set_defaults(action=("single", main_single))
 
     parser_query = subparsers.add_parser(
```


Longer term than this PR, I think we need to better handle invalid optional params in subcommands.

When subcommands are provided

Reply:
filed https://github.com/canonical/cloud-init/issues/4609 to track the above issue with CLI separately